### DISCUSSION TITLE
Additional launch options for restrictions

### DISCUSSION THREAD

**Discussion by DroidU** [owncast/owncast#2587] [Ideas]
I think there would be a need for the following restrictive startup options so that Owncast can also be used by streaming providers:

- -video-passthrough-only: Only Video Passthrough can be used. This saves a lot of CPU resources -> several clients can start their Owncast on the same server.
- -max-connection: Maximum number of simultaneously connected clients. The bandwidth is not exceeded if a customer has too many viewers, because it is limited. With this, several Owncasts can be started on the same server.
- -max-bitrate: Max bitrate/stream

The essence of the matter is, among other things, that if the streaming providers can also use these options, more will appear on the Owncast Directory page, which will make the software more popular.

**Comment by gabek** [comment]
It sounds like you want a set of features that are only applicable to your specific use case. Having startup flags that forcefully restrict what people can do with the software is not something an individual or an organization would ever want. People who run Owncast are allowed to configure it any way they want, it's their software.

If you look at our [Product Definition](https://github.com/owncast/owncast/blob/develop/docs/product-definition.md#primary-goals) You can see "For individuals, not service providers", and using that document as a guide to determine what work to prioritize I'm afraid options like this don't align with our goals.

**Reply by remco-k** [reply]
Found this discussion while searching for the presence of a max viewers option in Owncast. I would like to vote for adding a max viewer option, somewhere in Owncast.

@gabek 
> You can see "For individuals, not service providers", and using that document as a guide to determine what work to prioritize I'm afraid options like this don't align with our goals.

Fair point from the provider and goals view. 
But here's 2 situations (see below) I'm trying to get a solution for. The solution for both of them is having a max viewers option. Not specifically through startup parameters, I don't really care where the setting is going to be. In the admin panel would be perfectly fine for me. 

I'm working for a public organization (founding, local radio station) as a volunteer and I'm currently test-streaming with Owncast. The stream is 24/7 live with a high and low quality stream. The high one running at 2,5Mbps. The stream is running 24/7. 

**1. Ensuring stream stability**
A current limit that causes the stream to not play (or not play smoothly) is bandwith (1Gbps) or hardware (CPU/Memory/IO). A simple calculation states that 1Gbps/2,5Mbps=409. (_that is, with 409 theoretical maximum viewers we run out of bandwith and every viewer starts having dropouts. Due to overhead in Ethernet and TCP/IP and other things that apply, such as hardware limitations, this theoretical maximum is way less, probably about 200-300_).
Therefore, it is plausible (_and that has already happened in the past 4 weeks we have been testing_), that a unexpected high number of viewers are connected to the hls stream and things start to break apart, which is really bad. I would like to be able to prevent that BEFORE that happens and therefore guarantee (as much as I can) quality for already connected viewers. New ones should not be able to connect.
To achive this I want to be able to enter a max number of viewers for what I know the server can handle. So that the server never gets into a state it can't handle and things go bad for viewers.

**2. Prevent exceeding data traffic limits and extra costs for hosted servers**
Our server is hosted in a datacenter. Data traffic limits applies. If we exceed them, we get a warning e-mail and at the end of the month a invoice and we need to pay for the extra data traffic, in some cases costs might exceed the monthly fee for the server itself. All that needs to be prevented, and being able to set a maximum number of viewers helps with that. I am aware that it does not _guarantee_ anything for the used data traffic, but (in my decades of experience with audio streaming servers) this has been a sufficient method to not go over data traffic limits. Don't get me wrong, it might still do that in certain situations. That is acceptable. But at the moment we just have to little options to prevent this.
To be clear: I'm not asking for Owncast to manage the monthly data traffic. I'm asking for a max viewers limit option, that allows _me_ to manage the monthly data traffic.


**What should happen if a new viewer tries to connect but the limit is reached?**
The stream should not start (and should be logged). This enables (web)players to be able and go forward to another source url/server, such as audio only, in our case, which has _way_ more capacity than the video stream.

**Reply by gabek** [reply]
Thanks for the detail on your use case. Unfortunately as it stands, and I'm sure this seems weird, but the viewer count is more of a best guess estimate and not a 100% always accurate number that we would be able to to make max viewer decisions on.  Why is this?

- If we counted a viewer as the number of people downloading video from Owncast it wouldn't count the people downloading from a S3 provider at all. Or if a CDN is placed in front of an Owncast server that one person would be seen as multiple people coming from multiple locations.
- If we counted a viewer was the number of people currently on the Owncast web interface then it wouldn't count the people watching via external players such as on Smart TVs or 3rd party video players.

Right now we separately count the people downloading video _and_ the people on the web interface to make sure everyone is counted and and combine them to make a best effort guess, and for most situations it's correct. But when you start adding a CDN it no longer becomes accurate.

There's a future issue filed for me to try to improve that at https://github.com/owncast/owncast/issues/2627.

In short, using the number of viewers to manage how much bandwidth you're using is probably not a good idea. There are better options, such as using a S3 provider to distribute your video so your Owncast doesn't have to. That's why that option is there.

**Reply by remco-k** [reply]
I did read somewhere about the view count issue. So in a way I could have expected view count to not be the proper way to go. 
Regarding your S3 idea: Thing is, I'm not in search for more bandwith (via S3/CDN) or options to be able to serve to more users than my Owncast instance can do by itself. That is a budget issue I can't solve.

Instead, I'm in search for a method that enables a standalone Owncast (_With no S3/CDN in front of it_) to stop accepting new viewers whenever "it" determines we've ran out of CPU/Memory/bandwith.

Sorry for me trying again. Just to be sure.
In case of:

1. A standalone Owncast server, which heavily relies on its own hardware and bandwith, and 
2. In our case the built-in Owncast webinterface isn't used at all, and
3. All viewers (web, Chromecast and Airplay) use the `https://<hostname>/hls/stream.m3u8` url using the `VideoJS` player.

Would the view count then still be unusable for that, or just a bad idea in general? Maybe some other option to go with?
Though offtopic in this discussion, I'm open to suggestions for solving/ensuring the stream stability under high load and preventing data traffic bills.

Thx!

**Reply by remco-k** [reply]
Or just to make a long story short: in this use case I would be perfectly fine with the current view counter `n` and being able to set a limit at max `m` viewers somewhere and then Owncast just stops serving video to any new viewer if `n>=m`.

**Reply by gabek** [reply]
> * In our case the built-in Owncast webinterface isn't used at all

So in your case, what is a viewer? I know that sounds like a stupid question because it's clear a viewer is one person watching the video. But to Owncast, what is a viewer? Is it one IP address? If so, then people using a CDN will show up from multiple IP addresses for a single viewer, making Owncast overcount, leading Owncast to stopping new viewers too early. If you're using S3 then it would count zero people because nobody is actually touching the Owncast server, leading to Owncast never stopping new viewers. Usually this is resolved by Owncast counting the number of people watching via the web interface, but in your case it's not used, so that fallback isn't available.

I know you're not using a CDN or S3, but I have to think about the big picture, and adding a feature that stops serving requests after a certain number of viewers, but Owncast is unable to know the number of viewers reliably, is just building a broken feature.

There are other things to keep in mind when it comes to bandwidth as well.

- Not all viewers are playing the same output quality, meaning some people are using up more of your bandwidth than others. Not all viewers are the same.
- Peak concurrent users doesn't imply anything about your monthly usage. You might only have a few viewers at a time, but sustained over an entire month. Blocking peak max users won't help with that.

**Reply by remco-k** [reply]
All points taken.
Probably the most important one: `but I have to think about the big picture...is just building a broken feature`, sounds about right to me to not go down this road. :)

I'm going to see if I can think of something else for fixing my first point: ensuring stream stability under high load.
